### PR TITLE
Add VSTS support

### PR DESCRIPTION
**Pull request by jycouet**
Hi, 
Here is the draft of VSTS support of Renovate 😄 Link to #571 

I updated the doc also to find what needs to be configured... Here is the main idea:
`renovate {repo} --token {token} --platform vsts --endpoint https://{instance}.visualstudio.com/DefaultCollection`

Let me know about next steps
Thx

TODO:
- [x] Implement plain markdown PR body in master branch ( #1082 )
- [x] Match VSTS exports with other platforms (i.e. fix failing test)
- [x] Unit test VSTS methods
- [x] Add issues for all deferred methods
- [x] Have a working version!

### CODE REVIEW

**Comment by rarkins** [comment]
@jycouet wow!

Could you also document - somewhere - how to "quick start" setting up vsts so I can help out directly now and in future?

For now I'll reply in #571 although we may want to shift discussions here if we need to start commenting on actual code.

**Comment by jycouet** [comment]
Hey @rarkins, 
About how to I added the an example of cmd line in:
- https://github.com/singapore/renovate/pull/1049/files#diff-76e731333fb756df3bff5ddb3b731c46
- `renovate {repo} --token {token} --platform vsts --endpoint https://{instance}.visualstudio.com/DefaultCollection`

To get the token it's here: 
- https://github.com/singapore/renovate/pull/1049/files#diff-0730bb7c2e8f9ea2438b52e419dd86c9 
- https://www.visualstudio.com/en-us/docs/integrate/get-started/authentication/pats
---
I switch also to #571 for general discussions :)

**Comment by rarkins** [comment]
image

Awesome!

**Comment by rarkins** [comment]
Running todo list:

- [ ] Resolve PR body markdown problem

**Comment by jycouet** [comment]
What is the plan?
Platforms should have a "bodyFormat" function? And Globally we have an object containing only data?
You want to work on it? Should I do something?
let me know 😄

**Comment by rarkins** [comment]
We have two places where we create PR bodies - for onboarding, and for upgrades. I think in both cases we are using a mix of HTML and Markdown and then converting it completely to HTML before passing to passing to existing GitHub and GitLab APIs.

The need for HTML has been for reasons like:

- Markdown does not support a "headless" table, i.e. with no heading row
- Difficulties embedding things like `code` inside dot points

Ideally we would keep platform-specifics in the platform APIs, e.g. in regular code I write "Pull Request" or "PR" and then do a search/replace within the GitLab library itself. But it's not worth making it 3x more complex just to try to maintain that perfect separation. Perhaps we'll end up with a "plain" version and a "rich" version of PRs and we can configure it so that VSTS is the only platform with "plain".

Specifically for this PR, I suggest we aim to simply make it tolerable enough to merge, and then improve after. e.g. even if it has HTML exposed throughout, if having it merged and supported is better than not at all, then we merge it as-is or with minimal improvements.

**Comment by jycouet** [comment]
Hi @rarkins 
I had a look and yes it's not that easy ^^.
In onboarding, it's a mix of md and html yes... I refactored a bit, tell me if you like it? (no more table).
I feel that it's quite clean like this.

For upgrades it's another story since it's with `handlebars` and `showdown`!
I tried to just disable `showdown` to see the markdown only... but there are no "\n" so it's also not good for vsts :(

I do think that it's simpler with a "simple" markdown like I started... But it's a bit less fancy (without table). And I don't know the support of Gitlab. I can try a bit on the upgrades... just let me know.

**Comment by rarkins** [comment]
@jycouet, thanks I'll take a look. I think the best thing is to revert to using plain markdown (not HTML) even if it means not looking quite as good. I can take over.

File: lib/config/definitions.js [hunk 1]
```diff
@@ -8,8 +8,8 @@ function template(name, subdir = 'default/') {
     path.resolve(__dirname, '../config/templates/', subdir, shortName),
     'utf8'
   );
-  // Strip off any trailing line break
-  return hbsContents.replace(/\n$/, '');
+  // Strip off any trailing line break (handle for Macs, Linux, Windows)
+  return hbsContents.replace(/\r?\n|\r/g, '');
```
Comment (rarkins) [inline lib/config/definitions.js]: Can you break this into a separate PR? I like to keep unrelated changes to a minimum

File: lib/workers/global/index.js [hunk 2]
```diff
@@ -50,5 +50,6 @@ function getRepositoryConfig(globalConfig, index) {
   });
   repoConfig.isGitHub = repoConfig.platform === 'github';
   repoConfig.isGitLab = repoConfig.platform === 'gitlab';
+  repoConfig.isGitvsts = repoConfig.platform === 'vsts';
```
Comment (rarkins) [inline lib/workers/global/index.js]: Perhaps this can just be `isVsts` ? i.e. not sure if we need to prefix vsts with git?

File: lib/workers/bodyFormat/index.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+ * @param {string} list of String 
+ * @param {boolean} With number or bullet points
+ */
+function addSubList(lStr, isWithNumber = false) {
+  const specialchar = isWithNumber ? '\n  1. ' : '\n  - ';
+  return specialchar + lStr.join(specialchar);
+}
+
+function getBody() {
+  return body;
```
Comment (rarkins) [inline lib/workers/bodyFormat/index.js]: I will incorporate these ideas into a separate refactoring PR I'm working on, so that body changes are not present in this PR eventually

File: package.json [hunk 4]
```diff
@@ -72,7 +73,8 @@
     "showdown": "1.8.1",
     "tmp-promise": "1.0.3",
     "traverse": "0.6.6",
-    "yarn": "1.2.1"
+    "yarn": "1.2.1",
+    "vso-node-api": "6.2.8-preview"
```
Comment (rarkins) [inline package.json]: I assume this implies the API is not stable? But perhaps we don't have any other choice?

**Review by rarkins — CHANGES_REQUESTED** [review]
Also please remove the package-lock.json as we are using yarn.lock

File: lib/workers/bodyFormat/index.js (hunk 3 above)
Comment (jycouet) [reply lib/workers/bodyFormat/index.js]: Sure, I let you look into it :)

File: lib/config/definitions.js (hunk 1 above)
Comment (jycouet) [reply lib/config/definitions.js]: Yes, I'll look at it

File: lib/workers/global/index.js (hunk 2 above)
Comment (jycouet) [reply lib/workers/global/index.js]: On this one, yes I would like to keep isGitvsts since vsts supports also SVN and another one that I can't remember the name

File: package.json (hunk 4 above)
Comment (jycouet) [reply package.json]: yes, but it's the simplest was to interact with vsts. :(

**Comment by jycouet** [comment]
Hi, 
Could you have a look please?

Actually the `return hbsContents.replace(/\n$/, '');` is an issue with windows os.
And my previous "fix" was better, but not correct I think.
Since you are reworking markdown, you will remove it? or?
If not we will have to have a correct fix in another PR :)

File: lib/config/cli.js [hunk 5]
```diff
@@ -49,6 +49,9 @@ function getConfig(argv) {
       '    $ renovate --labels=renovate,dependency --ignore-unstable=false --log-level verbose singapore/lint-condo'
     );
     console.log('    $ renovate singapore/lint-condo singapore/package-test');
+    console.log(
+      '    $ renovate {repo} --token {token} --platform vsts --endpoint https://{instance}.visualstudio.com/DefaultCollection'
```
Comment (rarkins) [inline lib/config/cli.js]: The others here are *Examples* while this is kind of like a template. We should make them consistent

**Comment by rarkins** [comment]
@jycouet I've done some updates and now added a TODO list to the top comment/description in this PR

File: lib/config/index.js [hunk 6]
```diff
@@ -83,6 +89,12 @@ async function parseConfigs(env, argv) {
         config.token,
         config.endpoint
       );
+    } else if (config.platform === 'vsts') {
+      logger.info('Autodiscovering vsts repositories');
+      config.repositories = await vstsApi.getRepos(
```
Comment (rarkins) [inline lib/config/index.js]: This is not currently implemented. Is it possible? Or should we make a future issue for it even if so?

Comment (jycouet) [reply lib/config/index.js]: Done

File: lib/config/cli.js (hunk 5 above)
Comment (jycouet) [reply lib/config/cli.js]: I agree, just the problem is that there is no public endpoint, so I cannot provide an example.
You could put something like: `console.log('  Examples/templates:');` at the begining?

**Comment by jycouet** [comment]
Hey, 
- getRepos is now working
- I added all exports with a log info to say that it's not supported by VSTS (yet!)
  - Like this we can add one by one when needed? What do you think?
  - I left `getFile()` but I think it should be removed from exports on each platform since you use now `getFileContent`

**Comment by jycouet** [comment]
Today renovate tests are not working on my side (I guess that it's due do Windows and handlebars)
I think that it will be solved by body in markdown
I'll write tests as soon as we have bodies in markdown :)

Should I create a PR with what I started on body in Markdown? Or you had a look and you will do it? 
Let me know

Thx

**Comment by rarkins** [comment]
@jycouet FYI, I'm working on a solution to #1082 now. Once that's ready, we can solidify the API which will then leave just unit tests to complete

File: lib/platform/vsts/index.js [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+module.exports = {
+  // Initialization
+  getRepos,
+  initRepo,
+  setBaseBranch,
+  // Search
+  getFileList,
+  // Branch
+  branchExists,
+  getAllRenovateBranches, // Not implemented yet
```
Comment (rarkins) [inline lib/platform/vsts/index.js]: This is OK, it just means Renovate won't clean up its branches automatically if you leave them open or reconfigure something.

File: lib/platform/vsts/index.js [hunk 8]
```diff
@@ -0,0 +1,413 @@
...
+  isBranchStale, // Not implemented yet
```
Comment (rarkins) [inline lib/platform/vsts/index.js]: Renovate then won't rebase branches once they become stale (= behind master branch)

File: lib/platform/vsts/index.js [hunk 9]
```diff
@@ -0,0 +1,413 @@
...
+  getBranchPr,
+  getBranchStatus,
+  getBranchStatusCheck,
+  setBranchStatus,
+  deleteBranch, // Not implemented yet
```
Comment (rarkins) [inline lib/platform/vsts/index.js]: Mostly just important for cleanup, but should be a fairly simple implementation?

File: lib/platform/vsts/index.js [hunk 10]
```diff
@@ -0,0 +1,413 @@
...
+  mergeBranch, // Not implemented yet
```
Comment (rarkins) [inline lib/platform/vsts/index.js]: Only needed if branch merge is used (advanced feature), so OK to defer

File: lib/platform/vsts/index.js [hunk 11]
```diff
@@ -0,0 +1,413 @@
...
+  getBranchLastCommitTime, // Not implemented yet
```
Comment (rarkins) [inline lib/platform/vsts/index.js]: This is used to accurately time out branches if they remain stuck on "pending", OK to defer

File: lib/platform/vsts/index.js [hunk 12]
```diff
@@ -0,0 +1,413 @@
...
+  // issue
+  addAssignees, // Not implemented yet
+  addReviewers, // Not implemented yet
```
Comment (rarkins) [inline lib/platform/vsts/index.js]: These 2 are useful but not essential right away

File: lib/platform/vsts/index.js [hunk 13]
```diff
@@ -0,0 +1,413 @@
...
+  // Comments
+  ensureComment, // Not working
```
Comment (rarkins) [inline lib/platform/vsts/index.js]: Not essential - can be deferred

File: lib/platform/vsts/index.js [hunk 14]
```diff
@@ -0,0 +1,413 @@
...
+  ensureCommentRemoval, // Not implemented yet
+  // PR
+  findPr, // Done, But renovate rely on github object, I added a convertor
+  createPr,
+  getPr,
+  updatePr,
+  mergePr, // Not implemented yet
```
Comment (rarkins) [inline lib/platform/vsts/index.js]: Means automerge is not supported

**Comment by rarkins** [comment]
@jycouet updated this branch + added comments

**Comment by jycouet** [comment]
Thx @rarkins, I'll work on it tonight :)

**Comment by jycouet** [comment]
Could you help me on 2 thing?

- [ ] Tests on windows
When I run all tests locally I get a lots of error:
image
And here is the diff in the snapshot:
image
It's a problem of `\r\n` in the file & windows.
I can't find where you fill: `platform.createPr.mock.calls[0][2]`? to see if we can do something there so that tests are passing on Windows too.

- [ ] package.json on Windows
I'm not committing this file, but to have commands running, I need to do:
image

Any ideas on how to address these points? (sorry if it's not the right place, I could create an issue?)

**Comment by rarkins** [comment]
> tests on windows

Ideally we should get them all working cross-platform, although I'm a little lost on where to start. I'm not sure what you mean "I can't find where you fill: platform.createPr.mock.calls[0][2]" ? Anything `.mock.*` comes from Jest's mocking of those functions. How many do you have failing? Maybe using a git bash prompt could help?

> package.json on windows

probably best for a separate issue on Windows compatibility. Perhaps this might be a good solution: https://www.npmjs.com/package/cross-env

**Comment by jycouet** [comment]
Hey, 
I have 21 failing tests only about the `\r\n`.
I tested with a bach cmd but I get the same result.

In lib\workers\pr\index.js, If I add the line 172, I'm going to 20 fails! Maybe you could test in your machine to.
```
try {
   prBody = prBody.replace(/\r$/, ""); // <---- Line to add?!
   pr = await platform.createPr(branchName, prTitle, prBody, config.labels);
} catch (err) {
```

Anyway, let's open a new topic around Windows :)
BTW, cool lib cross-env

**Comment by jycouet** [comment]
Hi @rarkins, just to let you know that I'll try to finish this week :)
I'll let you know when on my side it's ok to merge

**Comment by rarkins** [comment]
It would be good to know where exactly the Windows line returns are coming from. Eg Are the \r being added by line returns in handlebars compiles?

**Comment by jycouet** [comment]
I think that yes, it's when handlebars compiles.
Look at the logs in trace level, we can see an `r`
image

**Comment by jycouet** [comment]
Could you check why prettier is failling on circleci? On my side it's working.
Could you review the status of the current PR? :)

thx

**Comment by rarkins** [comment]
@jycouet `prettier` was updated with some new rules recently, I'm guessing you maybe need to run `yarn install` again locally to get the latest package installed before you run `npm run prettier`. Anyway I have done this from my side and pushed to your branch now

**Comment by rarkins** [comment]
Re: the `\r`, actually that screenshot is helpful because it's showing that it is happening *before* handlebars compiles anything. It looks like the `\r` are being added when we do the "read templates from file" step so I've found a place to remove the trailing `\r` there. Can you see if that improves the number of tests passing for your locally? I'm not sure if we need to search/replace the `\r` *inside* the templates too (e.g. between lines, not just at the end of last line)/

**Comment by jycouet** [comment]
Hi @rarkins, you refactored error management not a long time ago right?
Now I get some `Repository has unknown error`, but it's hard to find where... I'm missing something?

**Comment by jycouet** [comment]
Now I have only 5 tests failing (locally on windows) due to `path.join()`
image
I did mock `path.join()` to see and all tests are passing! But when I mock this, jest is crashing at the end to generate the report!

Now I have this error when I launch renovate on a repo:
image
I saw that you reworked a lot... so you might know better where to look...

What should be the next step? :)

**Review by rarkins — APPROVED** [review]

**Comment by rarkins** [comment]
@jycouet perhaps we can use `path.normalize` to fix the `path.join` problem (see https://stackoverflow.com/questions/34329149/nodejs-absolute-paths-in-windows-with-forward-slash)

**Comment by jycouet** [comment]
Tadaaaaaa! 
I think we are in a good state @rarkins 

Anything to comment? add? remove?

**Comment by rarkins** [comment]
@jycouet shipped in v10.13.0! thanks so much for all your work. I also added you to the "contributors" field in Renovate's `package.json`

**Comment by jycouet** [comment]
Thank you @rarkins :)
